### PR TITLE
solution to "FinalLocalVariable doesn't report variable when first as…

### PR DESCRIPTION
**Pull request by Bhavik3**
…signment is separated" #1551

### CODE REVIEW

**Comment by romani** [comment]
Please review failures at orekit project, confirm that all of them is correct and please do PR to that fork.

**Comment by Bhavik3** [comment]
Done

**Comment by Bhavik3** [comment]
Done

**Comment by romani** [comment]
sorry for delay.

while I am doing code review please split commit into two with update to Check+Test and updates in checkstyle code for new functionality.

**Comment by romani** [comment]
FinalLocalVariableCheck.java

1) 
> An example of how to configure the check is:

please make it as "An example of how to configure the check to validate variable definition is:"

2)
> &lt;property name="token" value="VARIABLE_DEF"/&gt;

mismatch of documentation and getRequiredTokens(). please update all examples (javadoc, xdoc).

3)
>
```
    private final Deque<Map<String, DetailAST>> scopeStack = new ArrayDeque<>();

    private Deque<DetailAST> unInitializedVariable = new ArrayDeque<>();
```

please create new class that will store all information about scope (Map<String, DetailAST> + Deque<DetailAST>). 

4) Please create UT that show that variables in different but nested scopes are named the same and one of them final another is not final and issue is reported to required variable.

5)
please generate reports against project to see that new violations are correct , please do check all of them your self.

6)
> getDefaultToekns() , get AcceptableTokes()

pelase move TokenTypes.VARIABLE_DEF to the end of list to let future developer easily see a difference between collections,

7)
>private boolean defAndInitInDifferentLoop(DetailAST ast1, DetailAST ast2) 

Why only Loops are checked there, that should be processing for any block structure that create extra scope (scope is "{  }").

**Comment by romani** [comment]
@Bhavik3 , do you have time to finish that PR ?

**Comment by Bhavik3** [comment]
@romani

Little bit busy right now.

Let me know if It is at high priority, I will try to finish it...

**Comment by romani** [comment]
it is not a high priority, just try to finish all PRs that are hanging. 
Can you get rough estimation when you can come back to this PR ?

**Comment by Bhavik3** [comment]
maybe a week . . .

**Comment by romani** [comment]
Do you still plan to finish this PR ?

**Comment by Bhavik3** [comment]
@romani 
My health is not well for some days. I have dengue fever. I am admitted to hospital right now.
Really sorry for delay.

**Comment by romani** [comment]
@Bhavik3 , I sorry to hear this, take care of your health, opensource will wait for you.

I presume that after recovering you will not have time to finish this issue as you will need extra time to comeup at work and university. Most likely you will not be able to finish that PR till end on September.

I propose to you to drop this PR , and let other contributor to finish it. You will always be welcome at time with new contribution to our project. The only drawback of dropping is that you might loose authorship of changes, as we will need to rebase and additionally change your commit. Are you ok to drop this PR ?

**Comment by Bhavik3** [comment]
@romani 
you are right. I also miss my Mid-sem examination. so After recovery also, It will take some time to finish this PR. You can assign it to someone. He can continue my work.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 1]
```diff
@@ -274,6 +277,28 @@ private static boolean inLambda(DetailAST paramDef) {
     }
 
     /**
+     * if definition and initialization are in different for-loop, it can not be declared
+     * final. In that case, it will return false.
+     * @param ast1 VARIABLE_DEF
+     * @param ast2 init variable
+     * @return true if it is initiaized only once
+     */
+    private boolean defAndInitInDifferentLoop(DetailAST ast1, DetailAST ast2) {
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: Codacy Issue found: [The method defAndInitInDifferentLoop() has an NPath complexity of 48](https://www.codacy.com/app/checkstyle/checkstyle/file/1022207894/issues/source?bid=2193633&fileBranchId=2872415#l286)

**Comment by romani** [comment]
Fix discussion is moved to #2323

**Pull request closed without merging**